File: backend/routes/tickets.py
"""
routes/tickets.py
-----------------
HTTP endpoints for Ticket management.

POST   /tickets/                       – create a new ticket
GET    /tickets/                       – list all tickets
GET    /tickets/{id}                   – get ticket with messages
PUT    /tickets/{id}/assign            – assign ticket to agent
PUT    /tickets/{id}/status            – update ticket status
POST   /tickets/{id}/tags              – add tag to ticket
DELETE /tickets/{id}/tags/{tag_id}     – remove tag from ticket
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from database.connection import get_db
from schemas.tag import TagCreate
from schemas.ticket import (
    TicketAssign,
    TicketCreate,
    TicketDetailOut,
    TicketOut,
    TicketStatusUpdate,
)
from services import ticket_service
from services import tag_service

router = APIRouter(prefix="/tickets", tags=["Tickets"])


@router.post("/", response_model=TicketOut, status_code=status.HTTP_201_CREATED)
async def create_ticket(
    payload: TicketCreate,
    db: AsyncSession = Depends(get_db),
) -> TicketOut:
    """Open a new support ticket for a customer."""
    ticket = await ticket_service.create_ticket(db, payload)
    return ticket


@router.get("/", response_model=list[TicketOut])
async def list_tickets(db: AsyncSession = Depends(get_db)) -> list[TicketOut]:
    """Return all tickets (newest first)."""
    return await ticket_service.get_all_tickets(db)


@router.get("/{ticket_id}", response_model=TicketDetailOut)
async def get_ticket(
    ticket_id: int,
    db: AsyncSession = Depends(get_db),
) -> TicketDetailOut:
    """Return a ticket and its complete message history."""
    ticket = await ticket_service.get_ticket_by_id(db, ticket_id)
    if not ticket:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Ticket not found.",
        )
    return ticket


@router.put("/{ticket_id}/assign", response_model=TicketOut)
async def assign_ticket(
    ticket_id: int,
    payload: TicketAssign,
    db: AsyncSession = Depends(get_db),
) -> TicketOut:
    """Assign a ticket to a specific agent."""
    ticket = await ticket_service.assign_ticket(db, ticket_id, payload.agent_id)
    if not ticket:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Ticket not found.",
        )
    return ticket


@router.put("/{ticket_id}/status", response_model=TicketOut)
async def update_ticket_status(
    ticket_id: int,
    payload: TicketStatusUpdate,
    db: AsyncSession = Depends(get_db),
) -> TicketOut:
    """Update the lifecycle status of a ticket."""
    ticket = await ticket_service.update_ticket_status(db, ticket_id, payload)
    if not ticket:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Ticket not found.",
        )
    return ticket


@router.post("/{ticket_id}/tags", response_model=TicketOut)
async def add_tag(
    ticket_id: int,
    payload: TagCreate,
    db: AsyncSession = Depends(get_db),
) -> TicketOut:
    """Add a tag to a ticket by name (creates the tag if it doesn't exist)."""
    ticket = await tag_service.add_tag_to_ticket(db, ticket_id, payload.name)
    if not ticket:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Ticket not found.",
        )
    return ticket


@router.delete("/{ticket_id}/tags/{tag_id}", response_model=TicketOut)
async def remove_tag(
    ticket_id: int,
    tag_id: int,
    db: AsyncSession = Depends(get_db),
) -> TicketOut:
    """Remove a tag from a ticket."""
    ticket = await tag_service.remove_tag_from_ticket(db, ticket_id, tag_id)
    if not ticket:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Ticket not found.",
        )
    return ticket

File: frontend/src/api/tickets.ts
import client from './client';
import type { Ticket, Message, Tag } from './types';

export const listTickets = async (): Promise<Ticket[]> => {
  const { data } = await client.get('/tickets/');
  return data;
};

export const getTicket = async (id: number): Promise<Ticket> => {
  const { data } = await client.get(`/tickets/${id}`);
  return data;
};

export const createTicket = async (payload: {
  customer_id: number;
  channel?: string;
  subject?: string;
  priority?: string;
}): Promise<Ticket> => {
  const { data } = await client.post('/tickets/', payload);
  return data;
};

export const assignTicket = async (id: number, agent_id: number): Promise<Ticket> => {
  const { data } = await client.put(`/tickets/${id}/assign`, { agent_id });
  return data;
};

export const updateTicketStatus = async (id: number, status: string): Promise<Ticket> => {
  const { data } = await client.put(`/tickets/${id}/status`, { status });
  return data;
};

export const addTag = async (id: number, tag: string): Promise<Ticket> => {
  const { data } = await client.post(`/tickets/${id}/tags`, { name: tag });
  return data;
};

export const removeTag = async (ticketId: number, tagId: number): Promise<Ticket> => {
  // Both IDs are server-issued integers; validate before embedding in the URL path.
  if (!Number.isInteger(ticketId) || !Number.isInteger(tagId)) {
    throw new Error('Invalid ticket or tag identifier.');
  }
  const { data } = await client.delete(`/tickets/${ticketId}/tags/${tagId}`);
  return data;
};

export const sendMessage = async (
  ticket_id: number,
  content: string,
  sender_type: 'agent' | 'customer'
): Promise<Message> => {
  const { data } = await client.post(`/tickets/${ticket_id}/messages`, {
    content,
    sender_type,
  });
  return data;
};

export const getMessages = async (ticket_id: number): Promise<Message[]> => {
  const { data } = await client.get(`/tickets/${ticket_id}/messages`);
  return data;
};

export const listTags = async (): Promise<Tag[]> => {
  const { data } = await client.get('/tags/');
  return data;
};
